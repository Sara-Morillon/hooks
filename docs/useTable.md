# useTable

`useTable` offers an easy way to apply local filters, sort and pagination to a set of local data.

_:information_source: Pagination is 0 based._

## Example

```tsx
import { useTable } from '@saramorillon/hooks'

interface IData {
  name: string
  age: number
}

const data: IData[] = [
  { name: 'Zola Ray', age: 30 },
  { name: 'Arlo Curtis', age: 32 },
]

function MyComponent() {
  const { rows, total, state, filter, sort, goTo, setLimit } = useTable<IData>(data)

  return (
    <>
      <table>
        <thead>
          <tr>
            <th>
              Name <SortToggle dir={state.sortDir.name} onSort={(dir) => sort('name', dir)} />
            </th>
            <th>
              Age <SortToggle dir={state.sortDir.age} onSort={(dir) => sort('age', dir)} />
            </th>
          </tr>

          <tr>
            <td>
              <input value={state.name} onChange={(e) => filter('name', e.target.value)} />
            </td>
            <td>
              <input type="number" value={state.age} onChange={(e) => filter('age', e.target.value)} />
            </td>
          </tr>
        </thead>

        <tbody>
          {rows.map((row) => (
            <tr>
              <td>{row.name}</td>
              <td>{row.age}</td>
            </tr>
          ))}
        </tbody>
      </table>

      <Pagination pagination={state.pagination} goTo={goTo} setLimit={setLimit} total={total} />
    </>
  )
}
```

## API

```typescript
useTable<T, F extends IFilter<T>>(data: T[], initialState?: IInitialState<T, F>, options?: ITableOptions<T, F>): ITable<T, F>
```

### Arguments

`data: T[]` - the data that will be filtered, sorted and paginated.

`initialState?: IInitialState<T, F>` - the initial state of the table.

`options?: ITableOptions<T, F>` - an object containing:

- `filterFunctions: IFilterFunctions<T, F>` - functions to use to filter the data.

### Returns

`ITable<T, F>` - an object containing:

- `rows: T[]` - the filtered, sorted and paginated rows.
- `total: number` - the total number of rows available (total of filtered rows).
- `state: IState<T, F>` - an object containing:
  - `filter: Partial<F>` - the filter state.
  - `sort: ISort<T>[]` - the sort state, as an ordered list.
  - `sortDir: { [key in keyof T]?: 'asc' | 'desc' }` - the sort state, as an unordered object.
  - `pagination: IPaginationInfo` - an object containing:
    - `index: number` - the page index.
    - `limit: number` - the page size.
- `filter: <K extends keyof T>(field: K, value: F[K]) => void` - a function to apply a filter to a field.
- `sort: (field: keyof T, dir?: 'asc' | 'desc') => void` - a function to apply a sort to a field.
- `goTo: (index: number) => void` - a function to change page index.
- `setLimit: (limit: number) => void` - a function to change page size.
