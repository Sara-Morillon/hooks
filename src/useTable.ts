import { useMemo } from 'react'
import { type IFilterFunctions, type IFilterState, type Unknown, useFilterState, useFilteredRows } from './useFilter.js'
import { type IPaginationState, usePaginatedRows, usePaginationState } from './usePaginate.js'
import { type ISortFunctions, type ISortState, useSortState, useSortedRows } from './useSort.js'

export type IState<T, F extends Unknown<T> = T> = {
  filter: IFilterState<T, F>['state']
  sort: ISortState<T, F>['state']
  sortDir: {
    [key in keyof F]?: 'asc' | 'desc'
  }
  pagination: IPaginationState['state']
}

export type IInitialState<T, F extends Unknown<T> = T> = Partial<Omit<IState<T, F>, 'sortDir'>>

export interface ITableState<T, F extends Unknown<T> = T> {
  state: IState<T, F>
  filter: IFilterState<T, F>['filter']
  sort: ISortState<T, F>['sort']
  goTo: IPaginationState['goTo']
  setLimit: IPaginationState['setLimit']
}

export interface ITableRows<T> {
  rows: T[]
  total: number
}

export interface ITable<T, F extends Unknown<T> = T> extends ITableState<T, F>, ITableRows<T> {}

export interface ITableOptions<T, F extends Unknown<T> = T> {
  filterFunctions?: IFilterFunctions<T, F>
  sortFunctions?: ISortFunctions<T, F>
}

export function useTableState<T, F extends Unknown<T> = T>(initialState?: IInitialState<T, F>): ITableState<T, F> {
  const { state: filterState, filter } = useFilterState<T, F>(initialState?.filter)
  const { state: sortState, sort } = useSortState<T, F>(initialState?.sort)
  const { state: paginationState, goTo, setLimit } = usePaginationState(initialState?.pagination)

  const state = useMemo(() => {
    const state: IState<T, F> = {
      filter: filterState,
      sort: sortState,
      sortDir: {},
      pagination: paginationState,
    }

    for (const sort of state.sort) {
      state.sortDir[sort.field] = sort.dir
    }

    return state
  }, [filterState, sortState, paginationState])

  return useMemo(() => ({ state, filter, sort, goTo, setLimit }), [state, filter, sort, goTo, setLimit])
}

export function useTableRows<T, F extends Unknown<T> = T>(
  data: T[],
  state?: IInitialState<T, F>,
  options?: ITableOptions<T, F>,
): ITableRows<T> {
  const filteredRows = useFilteredRows<T, F>(data, state?.filter, options?.filterFunctions)
  const sortedRows = useSortedRows<T, F>(filteredRows, state?.sort, options?.sortFunctions)
  const rows = usePaginatedRows(sortedRows, state?.pagination)

  return useMemo(() => ({ rows, total: filteredRows.length }), [rows, filteredRows])
}

export function useTable<T, F extends Unknown<T> = T>(
  data: T[],
  initialState?: IInitialState<T, F>,
  options?: ITableOptions<T, F>,
): ITable<T, F> {
  const { state, filter, sort, goTo, setLimit } = useTableState<T, F>(initialState)
  const { rows, total } = useTableRows(data, state, options)

  return useMemo(
    () => ({ rows, total, state, filter, sort, goTo, setLimit }),
    [rows, total, state, filter, sort, goTo, setLimit],
  )
}
