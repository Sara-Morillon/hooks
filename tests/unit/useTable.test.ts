import { act, renderHook } from '@testing-library/react'
import { type IInitialState, type ITableOptions, useTable, useTableRows, useTableState } from '../../src/useTable.js'
import { type IData, mockTableData } from '../mock.js'

describe('useTableState', () => {
  it('should return table state', () => {
    const { result } = renderHook(() => useTableState<IData>())
    act(() => result.current.filter('name', 'a'))
    act(() => result.current.sort('age', 'desc'))
    act(() => result.current.setLimit(5))
    act(() => result.current.goTo(2))
    expect(result.current.state).toEqual({
      filter: { name: 'a' },
      sort: [{ column: 'age', dir: 'desc' }],
      sortDir: { age: 'desc' },
      pagination: { index: 2, limit: 5 },
    })
  })
})

describe('useTableRows', () => {
  it('should return filtered, sorted and paginated rows', () => {
    const state: IInitialState<IData> = {
      filter: { name: 'a' },
      sort: [{ column: 'age', dir: 'desc' }],
      pagination: { index: 2, limit: 5 },
    }
    const options: ITableOptions<IData> = { filterFunctions: { name: (row, value) => row.name.includes(value) } }
    const { result } = renderHook(() => useTableRows(mockTableData(), state, options))
    expect(result.current.rows).toEqual([
      { name: 'Annabella Vargas', age: 37 },
      { name: 'Stefan Hart', age: 35 },
      { name: 'Jaxen Smith', age: 35 },
      { name: 'Kiara Nava', age: 34 },
      { name: 'Deandre Vu', age: 32 },
    ])
  })

  it('should return total filtered rows', () => {
    const state: IInitialState<IData> = {
      filter: { name: 'a' },
    }
    const options: ITableOptions<IData> = { filterFunctions: { name: (row, value) => row.name.includes(value) } }
    const { result } = renderHook(() => useTableRows(mockTableData(), state, options))
    expect(result.current.total).toBe(13)
  })
})

describe('useTable', () => {
  it('should return table state', () => {
    const { result } = renderHook(() => useTable(mockTableData()))
    act(() => result.current.filter('name', 'a'))
    act(() => result.current.sort('age', 'desc'))
    act(() => result.current.setLimit(5))
    act(() => result.current.goTo(2))
    expect(result.current.state).toEqual({
      filter: { name: 'a' },
      sort: [{ column: 'age', dir: 'desc' }],
      sortDir: { age: 'desc' },
      pagination: { index: 2, limit: 5 },
    })
  })

  it('should return filtered, sorted and paginated rows', () => {
    const options: ITableOptions<IData> = { filterFunctions: { name: (row, value) => row.name.includes(value) } }
    const { result } = renderHook(() => useTable(mockTableData(), options))
    act(() => result.current.filter('name', 'a'))
    act(() => result.current.sort('age', 'desc'))
    act(() => result.current.setLimit(5))
    act(() => result.current.goTo(2))
    expect(result.current.rows).toEqual([
      { name: 'Annabella Vargas', age: 37 },
      { name: 'Stefan Hart', age: 35 },
      { name: 'Jaxen Smith', age: 35 },
      { name: 'Kiara Nava', age: 34 },
      { name: 'Deandre Vu', age: 32 },
    ])
  })

  it('should return total filtered rows', () => {
    const options: ITableOptions<IData> = { filterFunctions: { name: (row, value) => row.name.includes(value) } }
    const { result } = renderHook(() => useTable(mockTableData(), options))
    act(() => result.current.filter('name', 'a'))
    expect(result.current.total).toBe(13)
  })
})
