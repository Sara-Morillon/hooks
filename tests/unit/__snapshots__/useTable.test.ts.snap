// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useTable > should return filtered, sorted and paginated rows 1`] = `
[
  {
    "age": 37,
    "name": "Annabella Vargas",
  },
  {
    "age": 35,
    "name": "Stefan Hart",
  },
  {
    "age": 35,
    "name": "Jaxen Smith",
  },
  {
    "age": 34,
    "name": "Kiara Nava",
  },
  {
    "age": 32,
    "name": "Deandre Vu",
  },
]
`;

exports[`useTableRows > should return filtered, sorted and paginated rows 1`] = `
[
  {
    "age": 37,
    "name": "Annabella Vargas",
  },
  {
    "age": 35,
    "name": "Stefan Hart",
  },
  {
    "age": 35,
    "name": "Jaxen Smith",
  },
  {
    "age": 34,
    "name": "Kiara Nava",
  },
  {
    "age": 32,
    "name": "Deandre Vu",
  },
]
`;
